### ISSUE TITLE
Modal navigation back not working

### ISSUE THREAD

**Issue by sthewissen** [TinyStuff/TinyMvvm#48]
Using 3.0.1, I have the following navigation pattern:

- My root is a `TabbedPage`, within the first tab sits a `NavigationPage` with a `Page`.
- From that `Page` I open a new Modal which is also a `NavigationPage` with `Page`.
- Pushing a new `Page` from that modally displayed `Page` and subsequently calling `Back` does nothing.

As far as I can tell, the system crashes on this line:

https://github.com/TinyStuff/TinyMvvm/blob/07e4e0975cdfa0cb691b364b64abadc5e7bc459f/src/TinyMvvm.Forms/NavigationHelper/FormsNavigationHelper.cs#L340

With a:

```
Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
```

Because we're in a Modal popup with its own NavigationStack I would expect the first thing `Back` does to be to pop the newly pushed page on the Modal navigation stack. If there is nothing left to go `Back` to on that stack, I assume the result should be to pop the entire modal.

Changing that specific line to the following at the very least fixes the back navigation in my specific case, but I don't quite oversee the consequences this has for the rest yet. Hoping you have a bit more insight into that :)

```csharp
var prevPage = Application.Current.MainPage.Navigation.ModalStack[0];
```

**Comment by dhindrik** [comment]
Have you tried the latest preview? There was a fix in a PR about modals. If that does not fix the problem I can investigate it further.

Thank you for creating the issue.

**Comment by sthewissen** [comment]
Still occurring on 4.0.1-pre4. If you want to check out a sample;

https://github.com/sthewissen/MVP/tree/updating-the-lot

Go to create a contribution and pick a contribution type from the list. It _should_ pop the page back after picking.

**Comment by dhindrik** [comment]
It feels like something has changed with navigation in Xamarin.Forms since I wrote that code because it behaves very strangely. But I will investigate more how to fix it.

**Comment by BoobalanK** [comment]
Hey @dhindrik , I was having a problem in BackAsync. The ReturningParameter is null.

After i debugged the Sample Shell app from Samples Folder. I can see the 0 index in the NavigationStack is null. This is the reason the ReturningParameter never gets it value.

Hope this shed some light

**Comment by dhindrik** [comment]
@BoobalanK are you using Shell or not?

**Comment by BoobalanK** [comment]
Yes @dhindrik Using Shell. i tested the shell sample only from this location https://github.com/TinyStuff/TinyMvvm/tree/4.0.1-pre4/src/Samples

**Comment by dhindrik** [comment]
Are you using ShellViewBase?

**Comment by BoobalanK** [comment]
Yes @dhindrik

**Comment by vendhanparthy** [comment]
Hi @dhindrik, do you have any update for the ReturningParameter is null in BackAsync.

![image](https://user-images.githubusercontent.com/7837040/175981695-4e1886a7-1850-4d43-b90d-bc61fae604b5.png)
